openstack-api-charm-creation-guide.md

```diff
@@ -0,0 +1,240 @@
+
+### Create templates
+
+The charm code searches through the templates directories looking for a directory cooresponding to the Openstack release being installed
+or earlier. Since Liberty is the earliest release the charm is supporting a directory called liberty will house the templates.
+
+```mkdir -p templates/liberty
+cp <from pkg>/{api-paste.init,policy.json,congress.conf} templates/liberty```
+
+For the moment policy.json and api-paste.ini can be used without modification but congress.conf needs to be updated to be a template with site specific information as well as setting some constants. Taking the congress.conf add variables for the keystone and mysql config:
+
+```
+auth_strategy = keystone
+...
+# List of driver class paths to import. (list value)
+drivers = congress.datasources.neutronv2_driver.NeutronV2Driver,congress.datasources.glancev2_driver.GlanceV2Driver,congress.datasources.nova_driver.NovaDriver,congress.datasources.keystone_driver.KeystoneDriver,congress.datasources.ceilometer_driver.CeilometerDriver,congress.datasources.cinder_driver.CinderDriver,congress.datasources.swift_driver.SwiftDriver,congress.datasources.plexxi_driver.PlexxiDriver,congress.datasources.vCenter_driver.VCenterDriver,congress.datasources.murano_driver.MuranoDriver,congress.datasources.ironic_driver.IronicDriver
+...
+connection = {{ shared_db.uri }}
+...
+[keystone_authtoken]
+{% if identity_service.auth_host -%}
+auth_uri = {{ identity_service.service_protocol }}://{{ identity_service.service_host }}:{{ identity_service.service_port }}
+auth_url = {{ identity_service.auth_protocol }}://{{ identity_service.auth_host }}:{{ identity_service.auth_port }}
+auth_plugin = password
+project_domain_id = default
+user_domain_id = default
+project_name = {{ identity_service.service_tenant }}
+username = {{ identity_service.service_username }}
+password = {{ identity_service.service_password }}
+{% endif -%}
+
+```
+
+### Render the config
+
+Now the templates and interfaces are in place the configs can be rendered. A side-effect of rendering the configs is that any associated services are restarted. Finally, set the config.complete state this will be used later to trigger other events.
+
+```
+@reactive.when('shared-db.available')
+@reactive.when('identity-service.available')
+def render_congress_config(identity_interface, db_interface):
+    # Render the congress config files and communicate that by setting the
+    # config.complete state
+    charm = congress.CongressCharmFactory.charm(
+        interfaces=[identity_interface, db_interface]
+    )
+    charm.render_all_configs()
+    reactive.set_state('config.complete')
+```
+
+### Run DB Migration
+
+The DB migration can only be run once the config files are in place since as congress.conf will contain the DB connection information. To achieve this the DB migration is gated on the config.complete being set. Finally set the db.synched event so that this is only run once.
```

Comment:
Maybe reiterate which file to edit here, but by now it should really be obvious!